src/vellum/workflows/nodes/core/map_node/node.py

```diff
@@ -75,95 +75,85 @@ def run(self) -> Iterator[BaseOutput]:
+                                continue
+
+                            output_mapped_items[index] = output_value
+                            yield BaseOutput(
+                                name=output_reference.name,
+                                delta=(output_value, index, "FULFILLED"),
                             )
-                            continue
 
-                        output_mapped_items = mapped_items[output_reference.name]
-                        if index < 0 or index >= len(output_mapped_items):
-                            logger.error(
-                                "Invalid map node index", extra={"index": index, "output_name": output_reference.name}
-                            )
-                            continue
+                        fulfilled_iterations[index] = True
 
-                        output_mapped_items[index] = output_value
-                        yield BaseOutput(
-                            name=output_reference.name,
-                            delta=(output_value, index, "FULFILLED"),
+                    elif subworkflow_event.name == "workflow.execution.paused":
+                        raise NodeException(
+                            code=WorkflowErrorCode.INVALID_OUTPUTS,
+                            message=f"Subworkflow unexpectedly paused on iteration {index}",
+                        )
+                    elif subworkflow_event.name == "workflow.execution.rejected":
+                        raise NodeException(
+                            f"Subworkflow failed on iteration {index} with error: {subworkflow_event.error.message}",
+                            code=subworkflow_event.error.code,
                         )
+                except Empty:
+                    all_futures_done = all(future.done() for future in futures)
 
-                    fulfilled_iterations[index] = True
-                    if all(fulfilled_iterations):
-                        break
-
-                    if self.max_concurrency is not None:
-                        self._start_thread()
-                elif subworkflow_event.name == "workflow.execution.paused":
-                    raise NodeException(
-                        code=WorkflowErrorCode.INVALID_OUTPUTS,
-                        message=f"Subworkflow unexpectedly paused on iteration {index}",
-                    )
-                elif subworkflow_event.name == "workflow.execution.rejected":
-                    raise NodeException(
-                        f"Subworkflow failed on iteration {index} with error: {subworkflow_event.error.message}",
-                        code=subworkflow_event.error.code,
-                    )
-        except Empty:
-            pass
+                    if all_futures_done:
+                        if not all(fulfilled_iterations):
+                            if self._event_queue.empty():
+                                logger.warning("All threads completed but not all iterations fulfilled")
```

Comment:
[oos] we should likely add Rejected events for each iteration that is without one